{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c75a5a00",
   "metadata": {},
   "source": [
    "# Notebook for demonstrating evidence matching between assayed fusions and categorical fusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4ec41cf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from os import environ\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# These are the configurations for the gene normalizer and UTA databases. These should\n",
    "# be adjusted by the user based on the locations where these databases exist.\n",
    "environ[\"GENE_NORM_DB_URL\"] = \"postgresql://postgres@localhost:5432/gene_normalizer\"\n",
    "environ[\"UTA_DB_URL\"] = \"postgresql://uta_admin:uta@localhost:5432/uta/uta_20240523b\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c78c9c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from civicpy import civic\n",
    "\n",
    "from fusor.fusor import FUSOR\n",
    "from fusor.translator import Translator\n",
    "\n",
    "fusor = FUSOR()\n",
    "translator = Translator(fusor=fusor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "39397589",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate KIF5B::RET AssayedFusion from StarFusion file\n",
    "from pathlib import Path\n",
    "\n",
    "from cool_seq_tool.schemas import Assembly, CoordinateType\n",
    "\n",
    "from fusor.harvester import StarFusionHarvester\n",
    "\n",
    "path = Path(\"../../tests/fixtures/star-fusion.fusion_predictions.abridged.tsv\")\n",
    "harvester = StarFusionHarvester()\n",
    "fusions_list = harvester.load_records(path)\n",
    "assayed_fusion_star_fusion = await translator.from_star_fusion(\n",
    "        fusions_list[0],\n",
    "        CoordinateType.RESIDUE.value,\n",
    "        Assembly.GRCH38.value\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c4a0b16d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in accepted fusion variants\n",
    "variants = civic.get_all_fusion_variants(include_status=\"accepted\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4bd30b54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIF5B(entrez:3799)::RET(entrez:5979)\n",
      "CCDC6(entrez:8030)::RET(entrez:5979)\n",
      "v::RET(entrez:5979)\n",
      "KIF5B(entrez:3799)::EGFR(entrez:1956)\n"
     ]
    }
   ],
   "source": [
    "for fusion in variants:\n",
    "    if \"KIF5B\" in fusion.vicc_compliant_name or \"RET\" in fusion.vicc_compliant_name:\n",
    "        print(fusion.vicc_compliant_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f9bbef6",
   "metadata": {},
   "source": [
    "The output above lists all possible categorical fusions with KIF5B or RET as a partner. \n",
    "We expect matches for the first fusion as its partners are both KIF5B and RET (and have\n",
    "equivalent breakpoint locations), and the third fusion, as its 5' partner is a multiple \n",
    "possible genes element and its 3' partner is RET."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9b2d9760",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:cool_seq_tool.mappers.exon_genomic_coords:48406078 on NC_000023.11 does not occur within the exons for NM_005636.4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'type': <FUSORTypes.CATEGORICAL_FUSION: 'CategoricalFusion'>, 'structure': [{'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>, 'transcript': 'refseq:NM_004521.3', 'exonEnd': 24, 'exonEndOffset': 0, 'gene': {'conceptType': 'Gene', 'name': 'KIF5B', 'primaryCoding': {'id': 'hgnc:6324', 'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/', 'code': 'HGNC:6324'}}, 'elementGenomicEnd': {'id': 'ga4gh:SL.nk8wv9yKzCFQ0n7Ph2JnJhOkf2Fzfh_U', 'type': 'SequenceLocation', 'digest': 'nk8wv9yKzCFQ0n7Ph2JnJhOkf2Fzfh_U', 'sequenceReference': {'id': 'refseq:NC_000010.11', 'type': 'SequenceReference', 'refgetAccession': 'SQ.ss8r_wB0-b9r44TQTMmVTI92884QvBiB'}, 'start': 32017142}}, {'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>, 'transcript': 'refseq:NM_020975.6', 'exonStart': 11, 'exonStartOffset': 0, 'gene': {'conceptType': 'Gene', 'name': 'RET', 'primaryCoding': {'id': 'hgnc:9967', 'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/', 'code': 'HGNC:9967'}}, 'elementGenomicStart': {'id': 'ga4gh:SL.0VPUcnlmNOg_8HrscpHFr3HD2BNhHU3B', 'type': 'SequenceLocation', 'digest': '0VPUcnlmNOg_8HrscpHFr3HD2BNhHU3B', 'sequenceReference': {'id': 'refseq:NC_000010.11', 'type': 'SequenceReference', 'refgetAccession': 'SQ.ss8r_wB0-b9r44TQTMmVTI92884QvBiB'}, 'start': 43114479}}], 'viccNomenclature': 'NM_004521.3(KIF5B):e.24::NM_020975.6(RET):e.11', 'civicMolecularProfiles': [<CIViC molecular_profile 269>]}\n",
      "Match Score: 10\n",
      "{'type': <FUSORTypes.CATEGORICAL_FUSION: 'CategoricalFusion'>, 'structure': [{'type': <FUSORTypes.MULTIPLE_POSSIBLE_GENES_ELEMENT: 'MultiplePossibleGenesElement'>}, {'type': <FUSORTypes.GENE_ELEMENT: 'GeneElement'>, 'gene': {'conceptType': 'Gene', 'name': 'RET', 'primaryCoding': {'id': 'hgnc:9967', 'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/', 'code': 'HGNC:9967'}}}], 'viccNomenclature': 'v::RET(hgnc:9967)', 'civicMolecularProfiles': [<CIViC molecular_profile 1595>, <CIViC molecular_profile 4424>]}\n",
      "Match Score: 1\n"
     ]
    }
   ],
   "source": [
    "# Generate list of matches, report match score\n",
    "from fusor.fusion_matching import FusionMatcher\n",
    "from fusor.harvester import CIVICHarvester\n",
    "from fusor.models import FusionSet\n",
    "\n",
    "# Save categorical fusions cache and create FusionSet\n",
    "harvester = CIVICHarvester()\n",
    "harvester.fusions_list = variants\n",
    "fusions_list = harvester.load_records()\n",
    "\n",
    "civic_fusions = []\n",
    "for fusion in fusions_list:\n",
    "    if \"?\" in fusion.vicc_compliant_name:\n",
    "        continue\n",
    "    cex = await translator.from_civic(civic=fusion)\n",
    "    civic_fusions.append(cex)\n",
    "cache_dir = Path(\"../../src/fusor/data\")\n",
    "fs = FusionSet(assayedFusions=[assayed_fusion_star_fusion], categoricalFusions=civic_fusions)\n",
    "\n",
    "# Initialize FusionMatcher and define sources to match against\n",
    "fm = FusionMatcher(cache_dir=cache_dir,fusion_set=fs)\n",
    "\n",
    "# Generate list of matching fusions\n",
    "matches = await fm.match_fusion()\n",
    "for match in matches:\n",
    "    print(match[0].model_dump(exclude_none=True))\n",
    "    print(f\"Match Score: {match[1]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f048ccf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'type': <FUSORTypes.CATEGORICAL_FUSION: 'CategoricalFusion'>,\n",
       " 'structure': [{'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>,\n",
       "   'transcript': 'refseq:NM_004521.3',\n",
       "   'exonEnd': 24,\n",
       "   'exonEndOffset': 0,\n",
       "   'gene': {'conceptType': 'Gene',\n",
       "    'name': 'KIF5B',\n",
       "    'primaryCoding': {'id': 'hgnc:6324',\n",
       "     'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "     'code': 'HGNC:6324'}},\n",
       "   'elementGenomicEnd': {'id': 'ga4gh:SL.nk8wv9yKzCFQ0n7Ph2JnJhOkf2Fzfh_U',\n",
       "    'type': 'SequenceLocation',\n",
       "    'digest': 'nk8wv9yKzCFQ0n7Ph2JnJhOkf2Fzfh_U',\n",
       "    'sequenceReference': {'id': 'refseq:NC_000010.11',\n",
       "     'type': 'SequenceReference',\n",
       "     'refgetAccession': 'SQ.ss8r_wB0-b9r44TQTMmVTI92884QvBiB'},\n",
       "    'start': 32017142}},\n",
       "  {'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>,\n",
       "   'transcript': 'refseq:NM_020975.6',\n",
       "   'exonStart': 11,\n",
       "   'exonStartOffset': 0,\n",
       "   'gene': {'conceptType': 'Gene',\n",
       "    'name': 'RET',\n",
       "    'primaryCoding': {'id': 'hgnc:9967',\n",
       "     'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "     'code': 'HGNC:9967'}},\n",
       "   'elementGenomicStart': {'id': 'ga4gh:SL.0VPUcnlmNOg_8HrscpHFr3HD2BNhHU3B',\n",
       "    'type': 'SequenceLocation',\n",
       "    'digest': '0VPUcnlmNOg_8HrscpHFr3HD2BNhHU3B',\n",
       "    'sequenceReference': {'id': 'refseq:NC_000010.11',\n",
       "     'type': 'SequenceReference',\n",
       "     'refgetAccession': 'SQ.ss8r_wB0-b9r44TQTMmVTI92884QvBiB'},\n",
       "    'start': 43114479}}],\n",
       " 'viccNomenclature': 'NM_004521.3(KIF5B):e.24::NM_020975.6(RET):e.11',\n",
       " 'civicMolecularProfiles': [<CIViC molecular_profile 269>]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matches[0][0].model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "64d08226",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'type': <FUSORTypes.CATEGORICAL_FUSION: 'CategoricalFusion'>,\n",
       " 'structure': [{'type': <FUSORTypes.MULTIPLE_POSSIBLE_GENES_ELEMENT: 'MultiplePossibleGenesElement'>},\n",
       "  {'type': <FUSORTypes.GENE_ELEMENT: 'GeneElement'>,\n",
       "   'gene': {'conceptType': 'Gene',\n",
       "    'name': 'RET',\n",
       "    'primaryCoding': {'id': 'hgnc:9967',\n",
       "     'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "     'code': 'HGNC:9967'}}}],\n",
       " 'viccNomenclature': 'v::RET(hgnc:9967)',\n",
       " 'civicMolecularProfiles': [<CIViC molecular_profile 1595>,\n",
       "  <CIViC molecular_profile 4424>]}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matches[1][0].model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09ad5834",
   "metadata": {},
   "source": [
    "Both matching categorical fusions are selected and are sorted in descending order \n",
    "by the degree of match."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "86954c92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'_assertions': [],\n",
       " '_therapies': [<CIViC therapy 117>],\n",
       " '_phenotypes': [],\n",
       " '_incomplete': {'therapies'},\n",
       " '_partial': False,\n",
       " 'type': 'evidence',\n",
       " 'id': 698,\n",
       " 'variant_origin': 'SOMATIC',\n",
       " 'therapy_interaction_type': None,\n",
       " 'therapy_ids': [117],\n",
       " 'status': 'accepted',\n",
       " 'source_id': 378,\n",
       " 'significance': 'SENSITIVITYRESPONSE',\n",
       " 'rating': 2,\n",
       " 'phenotype_ids': [],\n",
       " 'name': 'EID698',\n",
       " 'molecular_profile_id': 269,\n",
       " 'evidence_type': 'PREDICTIVE',\n",
       " 'evidence_level': 'C',\n",
       " 'evidence_direction': 'SUPPORTS',\n",
       " 'disease_id': 30,\n",
       " 'description': 'A case study of a patient with EGFR, KRAS, BRAF, HER2, ALK, ROS1 and MET negative adenocarcinoma of the lung. FISH analysis revealed a KIF5B-RET fusion. The RET inhibitor Vandetanib led to remission in the patient.',\n",
       " 'assertion_ids': [],\n",
       " '_include_status': ['accepted', 'submitted', 'rejected']}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# View evidence item linked to matched categorical fusion\n",
    "matches[0][0].civicMolecularProfiles[0].evidence_items[0].__dict__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c8e9f12c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'_assertions': [<CIViC assertion 78>],\n",
       " '_therapies': [<CIViC therapy 601>],\n",
       " '_phenotypes': [],\n",
       " '_incomplete': {'therapies'},\n",
       " '_partial': False,\n",
       " 'type': 'evidence',\n",
       " 'id': 8852,\n",
       " 'variant_origin': 'SOMATIC',\n",
       " 'therapy_interaction_type': None,\n",
       " 'therapy_ids': [601],\n",
       " 'status': 'accepted',\n",
       " 'source_id': 3693,\n",
       " 'significance': 'SENSITIVITYRESPONSE',\n",
       " 'rating': 4,\n",
       " 'phenotype_ids': [],\n",
       " 'name': 'EID8852',\n",
       " 'molecular_profile_id': 1595,\n",
       " 'evidence_type': 'PREDICTIVE',\n",
       " 'evidence_level': 'A',\n",
       " 'evidence_direction': 'SUPPORTS',\n",
       " 'disease_id': 16,\n",
       " 'description': 'In this phase 1/2 trial (NCT03157128), patients with RET altered thyroid cancers were enrolled to receive the highly selective RET inhibitor selpercatinib. Among 19 patients with previously treated RET fusion positive thyroid cancer, the percentage who had a response was 79% (95% CI, 54 to 94), and 1-year progression-free survival was 64% (95% CI, 37 to 82). Responses included 8 of 11 (73%) papillary thyroid cancers, 2 of 3 poorly differentiated thyroid cancers, 1 of 2 anaplastic thyroid cancers, and 1 of 1 Hurthle-cell thyroid cancer. The most common fusions were CCDC6::RET and NCOA4::RET seen in 47% and 32% of fusion cases respectively.',\n",
       " 'assertion_ids': [78],\n",
       " '_include_status': ['accepted', 'submitted', 'rejected']}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# View evidence item linked to matched categorical fusion\n",
    "matches[1][0].civicMolecularProfiles[0].evidence_items[0].__dict__"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
