{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8233a206",
   "metadata": {},
   "source": [
    "# Transcript Accession Analysis\n",
    "This notebook highlights an example BCR::ABL1 fusion that is described in the literature using older transcript accessions versions and how FUSOR can be leveraged to represent this data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee38c60a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import environ\n",
    "\n",
    "# These are the configurations for the UTA and SeqRepo databases. These should\n",
    "# be adjusted by the user based on the locations where these databases exist.\n",
    "environ[\"UTA_DB_URL\"] = \"postgresql://anonymous@localhost:5432/uta/uta_20241220\"\n",
    "environ[\"SEQREPO_ROOT_DIR\"] = \"/usr/local/share/seqrepo/2024-12-20\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4c33ddd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***Using Gene Database Endpoint: http://localhost:8000***\n"
     ]
    }
   ],
   "source": [
    "from fusor.fusor import FUSOR\n",
    "\n",
    "fusor = FUSOR()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f00b19dc",
   "metadata": {},
   "source": [
    "Suppose our example fusion event is reported in an article from 2018 using the following syntax:\n",
    "**NM_004327.3:c.3000::NM_005157.5:c.2000**.\n",
    "\n",
    "These transcript accessions have since been updated to reflect newer versions.\n",
    "\n",
    "Assume these coordinates are described with respect to the CDS start site, and that these are **residue** coordinates. We would like to describe this fusion using genomic coordinates from the **GRCh38** build."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce32205b",
   "metadata": {},
   "source": [
    "## A. Using the `quick_start` module\n",
    "Run the cell below to quickly generate the `AssayedFusion` object using the `quick_start` module:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "934c7199",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'structure': [{'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>,\n",
       "   'transcript': 'refseq:NM_004327.3',\n",
       "   'transcriptStatus': <TranscriptPriority.LONGEST_COMPATIBLE_REMAINING: 'longest_compatible_remaining'>,\n",
       "   'strand': <Strand.POSITIVE: 1>,\n",
       "   'exonEnd': 16,\n",
       "   'exonEndOffset': -12,\n",
       "   'gene': {'conceptType': 'Gene',\n",
       "    'name': 'BCR',\n",
       "    'primaryCoding': {'id': 'hgnc:1014',\n",
       "     'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "     'code': 'HGNC:1014'}},\n",
       "   'elementGenomicEnd': {'id': 'ga4gh:SL.tQkBo8Xr_nDIElZFWqsy44yXXLxjW5Rd',\n",
       "    'type': 'SequenceLocation',\n",
       "    'extensions': [{'name': 'is_exonic', 'value': True}],\n",
       "    'digest': 'tQkBo8Xr_nDIElZFWqsy44yXXLxjW5Rd',\n",
       "    'sequenceReference': {'id': 'refseq:NC_000022.11',\n",
       "     'type': 'SequenceReference',\n",
       "     'refgetAccession': 'SQ.7B7SHsmchAR0dFcDCuSFjJAo7tX87krQ'},\n",
       "    'end': 23295143}},\n",
       "  {'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>,\n",
       "   'transcript': 'refseq:NM_005157.5',\n",
       "   'transcriptStatus': <TranscriptPriority.LONGEST_COMPATIBLE_REMAINING: 'longest_compatible_remaining'>,\n",
       "   'strand': <Strand.POSITIVE: 1>,\n",
       "   'exonStart': 11,\n",
       "   'exonStartOffset': 321,\n",
       "   'gene': {'conceptType': 'Gene',\n",
       "    'name': 'ABL1',\n",
       "    'primaryCoding': {'id': 'hgnc:76',\n",
       "     'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "     'code': 'HGNC:76'}},\n",
       "   'elementGenomicStart': {'id': 'ga4gh:SL.4AalyLSqDQxfxmlLmx0T9dizutqExADN',\n",
       "    'type': 'SequenceLocation',\n",
       "    'extensions': [{'name': 'is_exonic', 'value': True}],\n",
       "    'digest': '4AalyLSqDQxfxmlLmx0T9dizutqExADN',\n",
       "    'sequenceReference': {'id': 'refseq:NC_000009.12',\n",
       "     'type': 'SequenceReference',\n",
       "     'refgetAccession': 'SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI'},\n",
       "    'start': 130884289}}],\n",
       " 'type': <FUSORTypes.ASSAYED_FUSION: 'AssayedFusion'>}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from fusor.quick_start import FUSORQuickStart\n",
    "\n",
    "fqs = FUSORQuickStart(fusor=fusor,\n",
    "                      five_prime_junction=\"NM_004327.3:c.3000\",\n",
    "                      three_prime_junction=\"NM_005157.5:c.2000\",\n",
    "                      five_prime_gene=\"BCR\",\n",
    "                      three_prime_gene=\"ABL1\",\n",
    "                      cds_start_site=True\n",
    ")\n",
    "fusion = await fqs.quick_start()\n",
    "fusion.model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5763d43",
   "metadata": {},
   "source": [
    "## B. Using FUSOR/Cool-Seq-Tool modules\n",
    "This section shows how the same object can be generated using individual function calls to FUSOR and Cool-Seq-Tool that are then combined together."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c42e0ea2",
   "metadata": {},
   "source": [
    "## Representing the 5' partner\n",
    "The following cells can be run to generate a `TranscriptSegmentElement` for the 5' partner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a660908",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The CDS start is: 596'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_range = await fusor.cool_seq_tool.uta_db.get_cds_start_end(tx_ac=\"NM_004327.3\")\n",
    "cds_start = cds_range[0]\n",
    "f\"The CDS start is: {cds_start}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "22287dd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene': 'BCR',\n",
       " 'strand': <Strand.POSITIVE: 1>,\n",
       " 'tx_pos_range': (3476, 3608),\n",
       " 'alt_pos_range': (23295023, 23295155),\n",
       " 'alt_aln_method': 'splign',\n",
       " 'tx_exon_id': 956565,\n",
       " 'alt_exon_id': 6619783,\n",
       " 'tx_ac': 'NM_004327.3',\n",
       " 'alt_ac': 'NC_000022.11',\n",
       " 'coding_start_site': 0,\n",
       " 'coding_end_site': 0,\n",
       " 'alt_pos_change_range': (23295142, 23295143),\n",
       " 'pos_change': (119, 12)}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from cool_seq_tool.schemas import Assembly\n",
    "\n",
    "tx_data = await fusor.cool_seq_tool.uta_db.get_genomic_tx_data(\n",
    "    tx_ac=\"NM_004327.3\",\n",
    "    pos=(cds_start+3000-1,cds_start+3000),\n",
    "    target_genome_assembly=Assembly.GRCH38\n",
    ")\n",
    "tx_data.model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fe4103a",
   "metadata": {},
   "source": [
    "We can now use the outputted data to create a `TranscriptSegmentElement` using FUSOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "92593108",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>,\n",
       " 'transcript': 'refseq:NM_004327.3',\n",
       " 'transcriptStatus': <TranscriptPriority.LONGEST_COMPATIBLE_REMAINING: 'longest_compatible_remaining'>,\n",
       " 'strand': <Strand.POSITIVE: 1>,\n",
       " 'exonEnd': 16,\n",
       " 'exonEndOffset': -12,\n",
       " 'gene': {'conceptType': 'Gene',\n",
       "  'name': 'BCR',\n",
       "  'primaryCoding': {'id': 'hgnc:1014',\n",
       "   'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "   'code': 'HGNC:1014'}},\n",
       " 'elementGenomicEnd': {'id': 'ga4gh:SL.tQkBo8Xr_nDIElZFWqsy44yXXLxjW5Rd',\n",
       "  'type': 'SequenceLocation',\n",
       "  'extensions': [{'name': 'is_exonic', 'value': True}],\n",
       "  'digest': 'tQkBo8Xr_nDIElZFWqsy44yXXLxjW5Rd',\n",
       "  'sequenceReference': {'id': 'refseq:NC_000022.11',\n",
       "   'type': 'SequenceReference',\n",
       "   'refgetAccession': 'SQ.7B7SHsmchAR0dFcDCuSFjJAo7tX87krQ'},\n",
       "  'end': 23295143}}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from cool_seq_tool.schemas import CoordinateType\n",
    "\n",
    "five_prime_transcript = await fusor.transcript_segment_element(\n",
    "    tx_to_genomic_coords=False,\n",
    "    transcript=\"NM_004327.3\",\n",
    "    genomic_ac=\"NC_000022.11\",\n",
    "    seg_end_genomic=23295143,\n",
    "    coordinate_type=CoordinateType.RESIDUE\n",
    ")\n",
    "five_prime_transcript = five_prime_transcript[0]\n",
    "five_prime_transcript.model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fefd57f",
   "metadata": {},
   "source": [
    "## Representing the 3' partner\n",
    "The following cells can be run to generate a `TranscriptSegmentElement` for the 3' partner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5107942c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The CDS start is: 192'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_range = await fusor.cool_seq_tool.uta_db.get_cds_start_end(tx_ac=\"NM_005157.5\")\n",
    "cds_start = cds_range[0]\n",
    "f\"The CDS start is: {cds_start}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d0e3f8f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene': 'ABL1',\n",
       " 'strand': <Strand.POSITIVE: 1>,\n",
       " 'tx_pos_range': (1870, 5577),\n",
       " 'alt_pos_range': (130883968, 130887675),\n",
       " 'alt_aln_method': 'splign',\n",
       " 'tx_exon_id': 3530423,\n",
       " 'alt_exon_id': 6630666,\n",
       " 'tx_ac': 'NM_005157.5',\n",
       " 'alt_ac': 'NC_000009.12',\n",
       " 'coding_start_site': 0,\n",
       " 'coding_end_site': 0,\n",
       " 'alt_pos_change_range': (130884289, 130884290),\n",
       " 'pos_change': (321, 3385)}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx_data = await fusor.cool_seq_tool.uta_db.get_genomic_tx_data(\n",
    "    tx_ac=\"NM_005157.5\",\n",
    "    pos=(cds_start+2000-1,cds_start+2000),\n",
    "    target_genome_assembly=Assembly.GRCH38\n",
    ")\n",
    "tx_data.model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aec40e11",
   "metadata": {},
   "source": [
    "We can now use the outputted data to create a `TranscriptSegmentElement` using FUSOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2ef714a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>,\n",
       " 'transcript': 'refseq:NM_005157.5',\n",
       " 'transcriptStatus': <TranscriptPriority.LONGEST_COMPATIBLE_REMAINING: 'longest_compatible_remaining'>,\n",
       " 'strand': <Strand.POSITIVE: 1>,\n",
       " 'exonStart': 11,\n",
       " 'exonStartOffset': 321,\n",
       " 'gene': {'conceptType': 'Gene',\n",
       "  'name': 'ABL1',\n",
       "  'primaryCoding': {'id': 'hgnc:76',\n",
       "   'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "   'code': 'HGNC:76'}},\n",
       " 'elementGenomicStart': {'id': 'ga4gh:SL.4AalyLSqDQxfxmlLmx0T9dizutqExADN',\n",
       "  'type': 'SequenceLocation',\n",
       "  'extensions': [{'name': 'is_exonic', 'value': True}],\n",
       "  'digest': '4AalyLSqDQxfxmlLmx0T9dizutqExADN',\n",
       "  'sequenceReference': {'id': 'refseq:NC_000009.12',\n",
       "   'type': 'SequenceReference',\n",
       "   'refgetAccession': 'SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI'},\n",
       "  'start': 130884289}}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "three_prime_transcript = await fusor.transcript_segment_element(\n",
    "    tx_to_genomic_coords=False,\n",
    "    transcript=\"NM_005157.5\",\n",
    "    genomic_ac=\"NC_000009.12\",\n",
    "    seg_start_genomic=130884290,\n",
    "    coordinate_type=CoordinateType.RESIDUE\n",
    ")\n",
    "three_prime_transcript = three_prime_transcript[0]\n",
    "three_prime_transcript.model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef593069",
   "metadata": {},
   "source": [
    "## Create `AssayedFusion` object \n",
    "We can now assemble the fusion event, described with FUSOR using the provided transcript accessions from the literature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aa83bc07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'structure': [{'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>,\n",
       "   'transcript': 'refseq:NM_004327.3',\n",
       "   'transcriptStatus': <TranscriptPriority.LONGEST_COMPATIBLE_REMAINING: 'longest_compatible_remaining'>,\n",
       "   'strand': <Strand.POSITIVE: 1>,\n",
       "   'exonEnd': 16,\n",
       "   'exonEndOffset': -12,\n",
       "   'gene': {'conceptType': 'Gene',\n",
       "    'name': 'BCR',\n",
       "    'primaryCoding': {'id': 'hgnc:1014',\n",
       "     'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "     'code': 'HGNC:1014'}},\n",
       "   'elementGenomicEnd': {'id': 'ga4gh:SL.tQkBo8Xr_nDIElZFWqsy44yXXLxjW5Rd',\n",
       "    'type': 'SequenceLocation',\n",
       "    'extensions': [{'name': 'is_exonic', 'value': True}],\n",
       "    'digest': 'tQkBo8Xr_nDIElZFWqsy44yXXLxjW5Rd',\n",
       "    'sequenceReference': {'id': 'refseq:NC_000022.11',\n",
       "     'type': 'SequenceReference',\n",
       "     'refgetAccession': 'SQ.7B7SHsmchAR0dFcDCuSFjJAo7tX87krQ'},\n",
       "    'end': 23295143}},\n",
       "  {'type': <FUSORTypes.TRANSCRIPT_SEGMENT_ELEMENT: 'TranscriptSegmentElement'>,\n",
       "   'transcript': 'refseq:NM_005157.5',\n",
       "   'transcriptStatus': <TranscriptPriority.LONGEST_COMPATIBLE_REMAINING: 'longest_compatible_remaining'>,\n",
       "   'strand': <Strand.POSITIVE: 1>,\n",
       "   'exonStart': 11,\n",
       "   'exonStartOffset': 321,\n",
       "   'gene': {'conceptType': 'Gene',\n",
       "    'name': 'ABL1',\n",
       "    'primaryCoding': {'id': 'hgnc:76',\n",
       "     'system': 'https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/',\n",
       "     'code': 'HGNC:76'}},\n",
       "   'elementGenomicStart': {'id': 'ga4gh:SL.4AalyLSqDQxfxmlLmx0T9dizutqExADN',\n",
       "    'type': 'SequenceLocation',\n",
       "    'extensions': [{'name': 'is_exonic', 'value': True}],\n",
       "    'digest': '4AalyLSqDQxfxmlLmx0T9dizutqExADN',\n",
       "    'sequenceReference': {'id': 'refseq:NC_000009.12',\n",
       "     'type': 'SequenceReference',\n",
       "     'refgetAccession': 'SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI'},\n",
       "    'start': 130884289}}],\n",
       " 'type': <FUSORTypes.ASSAYED_FUSION: 'AssayedFusion'>}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion = fusor.assayed_fusion(structure=[five_prime_transcript, three_prime_transcript])\n",
    "fusion.model_dump(exclude_none=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "fusor",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
